# **Welcome to _T-lex 3_ Github page**

**T-lex is a computational pipeline that detects presence and/or absence of annotated individual transposable elements (TEs) using next-generation sequencing (NGS) data. **


### Folder content
1. Manual.md : manual updated for T-lex 3
2. tlex-open-v3.1.pl : code file for running T-lex 3
3. example folder: contains an example file for running T-lex 3
4. Dockerfile : DOCKER file to run T-lex 3 as a container (not available for -_tsd_ option)
5. docker.md : manual for running T-lex 3 in DOCKER
6. supporting_information folder: contains scripts and files for calculating sensitivity and specificity


### How to cite _T-lex_ :
- María Bogaerts-Márquez, Maite G Barrón, Anna-Sophie Fiston-Lavier, Pol Vendrell-Mir, Raúl Castanera, Josep M Casacuberta, Josefa González, T-lex3: an accurate tool to genotype and estimate population frequencies of transposable elements using the latest short-read whole genome sequencing data, Bioinformatics, Volume 36, Issue 4, 15 February 2020, Pages 1191–1197, https://doi.org/10.1093/bioinformatics/btz727
- Fiston-Lavier AS, Barrón MG, Petrov DA, González J. T-lex2: genotyping, frequency estimation and re-annotation of transposable elements using single or pooled next-generation sequencing data. Nucleic Acids Res. 2015 Feb 27;43(4):e22. doi: 10.1093/nar/gku1250. Epub 2014 Dec 15. PMID: 25510498; PMCID: PMC4344482.
- Fiston-Lavier AS, Carrigan M, Petrov DA, González J. T-lex: a program for fast and accurate assessment of transposable element presence using next-generation sequencing data. Nucleic Acids Res. 2011 Mar;39(6):e36. doi: 10.1093/nar/gkq1291. Epub 2010 Dec 21. PMID: 21177644; PMCID: PMC3064797.
